network/src/protocols/rpc/mod.rs

```diff
@@ -337,6 +322,270 @@ impl InboundRpcs {
+            let err = Err(RpcError::TooManyPending(self.max_concurrent_outbound_rpcs));
+            let _ = application_response_tx.send(err);
+            return Err(RpcError::TooManyPending(self.max_concurrent_outbound_rpcs));
+        }
+
+        let request_id = self.request_id_gen.next();
+
+        trace!(
+            NetworkSchema::new(network_context).remote_peer(peer_id),
+            "{} Sending outbound rpc request with request_id {} and protocol_id {} to {}",
+            network_context,
+            request_id,
+            protocol_id,
+            peer_id.short_str(),
+        );
+
+        // Start timer to collect outbound RPC latency.
+        let timer =
+            counters::outbound_rpc_request_latency(network_context, protocol_id).start_timer();
+
+        // Enqueue rpc request message onto outbound write queue.
+        let message = NetworkMessage::RpcRequest(RpcRequest {
+            protocol_id,
+            request_id,
+            priority: Priority::default(),
+            raw_request: Vec::from(request_data.as_ref()),
+        });
+        let (ack_tx, _) = oneshot::channel();
+        write_reqs_tx.send((message, ack_tx)).await?;
+
+        // Collect counters for requests sent.
+        counters::rpc_messages(network_context, REQUEST_LABEL, SENT_LABEL).inc();
+        counters::rpc_bytes(network_context, REQUEST_LABEL, SENT_LABEL).inc_by(req_len);
+
+        // Create channel over which response is delivered to outbound_rpc_task.
+        let (response_tx, response_rx) = oneshot::channel::<RpcResponse>();
+
+        // Store send-side in the pending map so we can notify outbound_rpc_task
+        // when the rpc response has arrived.
+        self.pending_outbound_rpcs.insert(request_id, response_tx);
+
+        // A future that waits for the rpc response with a timeout. We create the
+        // timeout out here to start the timer as soon as we push onto the queue
+        // (as opposed to whenever it first gets polled on the queue).
+        let mut wait_for_response = tokio::time::timeout(timeout, response_rx).map(|result| {
+            // Flatten errors.
+            match result {
+                Ok(Ok(response)) => Ok(Bytes::from(response.raw_response)),
+                Ok(Err(oneshot::Canceled)) => Err(RpcError::UnexpectedResponseChannelCancel),
+                Err(_elapsed) => Err(RpcError::TimedOut),
+            }
+        });
+
+        // A future that waits for the response and sends it to the application.
+        let notify_application = async move {
+            let mut cancellation = application_response_tx.cancellation().fuse();
+            futures::select! {
+                maybe_response = wait_for_response => {
+                    // TODO(philiphayes): Clean up RpcError. Effectively need to
+                    // clone here to pass the result up to application layer, but
+                    // RpcError is not currently cloneable.
```

Comment:
Let's make a github issue to follow up about this :)

Reply:
https://github.com/diem/diem/issues/7229